Implement Algebra.Graph.Undirected

bolt12 opened the pull request:
I was based on the 'Graph.hs' file to get the most of what was done on the Graph data type. I kept some of the rewrite rules but will remove them if they do not make sense in this context.

This draft PR compiles fine. I'm looking for some reviews on what I've done so far.

There are a couple of things not working according to the `SymmetricRelation` API, more specifically the functions that involve transforming the graph to an `AdjacencyMap`. I think this should be fixed by implementing an `toUndirectedAdjacencyMap`! The problem is the original `toAdjacencyMap` is built with respect to `AdjacencyMap` which `connect` does not commute.

How should I deal with this case? Implement an entire interface called `UndirectedAdjacencyMap` or implement a version of the graph construction primitive in the `AdjacencyMap` module that preserve the 'Undirected' rules?

snowleopard commented on algebraic-graphs.cabal:
```diff
@@ -66,6 +66,7 @@ source-repository head
 library
     hs-source-dirs:     src
     exposed-modules:    Algebra.Graph,
+                        Algebra.Graph.Undirected.UndirectedGraph,
```
This should be just `Algebra.Graph.Undirected`.

snowleopard commented on algebraic-graphs.cabal:
```diff
@@ -66,6 +66,7 @@ source-repository head
 library
     hs-source-dirs:     src
     exposed-modules:    Algebra.Graph,
+                        Algebra.Graph.Undirected.UndirectedGraph,
```
This should be just `Algebra.Graph.Undirected`.

bolt12 replied:
Got it!

snowleopard commented on src/Algebra/Graph/Undirected/UndirectedGraph.hs:
```diff
@@ -0,0 +1,935 @@
+-----------------------------------------------------------------------------
+module Algebra.Graph.Undirected.UndirectedGraph (
+    -- * Algebraic data type for graphs
+    UndirectedGraph (..),
```
Let's call this data type simply `Graph`. Users will use qualified modules to disambiguate.

snowleopard replied:
Also, let's not export the constructor and see how an abstract implementation will work.

bolt12 replied:
okay!

snowleopard commented on src/Algebra/Graph/Undirected/UndirectedGraph.hs:
```diff
@@ -0,0 +1,935 @@
+newtype UndirectedGraph a = UndirectedGraph { graph :: G.Graph a }
+             deriving (Generic, NFData)
+
+instance Show a => Show (UndirectedGraph a) where
```
Let's implement the `Show` instance using your implementation in `Symmetric.Relation`.

snowleopard commented on src/Algebra/Graph/Undirected/UndirectedGraph.hs:
```diff
@@ -0,0 +1,935 @@
+instance Show a => Show (UndirectedGraph a) where
+  show = show . graph
+
+{- Note [Functions for rewrite rules]
```
Instead of copying this whole note, better say "See Note [Functions for rewrite rules] in Algebra.Graph.".

snowleopard commented on src/Algebra/Graph/Undirected/UndirectedGraph.hs:
```diff
@@ -0,0 +1,935 @@
+-- 'size'        empty == 1
+-- @
+empty :: UndirectedGraph a
+empty = UndirectedGraph G.Empty
```
Most of the functions in the API can be implemented using type-safe coercion, e.g. `empty = coerce G.empty`.

bolt12 replied:
Got it! :)

bolt12 replied:
How can I resolve this error?

![image](https://user-images.githubusercontent.com/20261421/61046436-093f0f80-a3d5-11e9-9952-edf98a326f85.png)

snowleopard commented on src/Algebra/Graph/Undirected/UndirectedGraph.hs:
```diff
@@ -0,0 +1,935 @@
+-- 'size'        (vertex x) == 1
+-- @
+vertex :: a -> UndirectedGraph a
+vertex = UndirectedGraph . G.Vertex
```
Same here: `vertex = coerce G.vertex`.

snowleopard commented:
@bolt12 Many thanks for the PR! I left a few comments.

Regarding `toAdjacencyMap`: Right now we do not have `Algebra.Graph.AdjacencyMap.Undirected`, but we do have `Algebra.Graph.Relation.Symmetric` -- let's use it instead, perhaps, adding a TODO that we should really switch to `Algebra.Graph.AdjacencyMap.Undirected` at some point in future (because it's typically faster and more convenient for algorithms).

bolt12 commented:
@snowleopard That's a really good idea! I've already resolved some of the points you mentioned. I left some more TODO's regarding the 'Algebra.Graph.AdjacencyMap.Undirected' but maybe I'll open an issue about this to register it.

Thanks for your reviews! I had problems on the coercion of the `empty` function and I still have problems when running ghci on the project root, but when i run `ghci src/Algebra/Graph/Undirected.hs` it compiles the module just fine!

![image](https://user-images.githubusercontent.com/20261421/61047844-52dd2980-a3d8-11e9-8857-7ff83d987a95.png)

Should I open an issue about this too?

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,907 @@
+{-# LANGUAGE DeriveGeneric, RankNTypes #-}
+-----------------------------------------------------------------------------
+-- |
+-- Module     : Algebra.Graph.Undirected.UndirectedGraph
```
Still old name.

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,907 @@
+'edge' 1 1 < 'edge' 1 2
+'edge' 1 2 < 'edge' 1 1 + 'edge' 2 2
+'edge' 1 2 < 'edge' 1 3@
+'edge' 1 2 == 'edge' 2 1
```
Misplaced `@`.

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,907 @@
+x     <= x + y
+x + y <= x * y@
+-}
+newtype Graph a = UndirectedGraph { graph :: G.Graph a }
```
We don't want to export the constructor, so let's give it a short name `UG`. Also, I don't think we need `graph`.

bolt12 replied:
Got it!

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,907 @@
+
+-- TODO: Find a more efficient equality check.
+-- TODO: Implement toUndirectedAdjacencyMap.
+-- Check if two graphs are equal by converting them to their adjacency maps.
```
Outdated comment.

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,907 @@
+
+-- Like 'eqR' but specialised for graphs with vertices of type 'Int'.
+eqIntR :: Graph Int -> Graph Int -> Bool
+eqIntR x y = toSymmetricRelation x == toSymmetricRelation y
```
At the moment this function is not specialised at all, since we don't have have `Relation` specialised to vertices of type `Int`. Let's keep the code as is, but add a `TODO`, saying that this needs to be fixed in future.

bolt12 replied:
Yes, I left it there and meant to leave a note saying that it wasn't really specialized, but forgot! I will fix that!

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,907 @@
+
+-- Like 'ordR' but specialised for graphs with vertices of type 'Int'.
+ordIntR :: Graph Int -> Graph Int -> Ordering
+ordIntR x y = compare (toSymmetricRelation x) (toSymmetricRelation y)
```
Same here: this is not really specialised.

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,907 @@
+-- 'size'        empty == 1
+-- @
+empty :: Graph a
+empty = UndirectedGraph G.empty
```
So, you didn't manage to write `coerce G.empty` here?

bolt12 replied:
No, GHC gave me the error I posted... It's something with the quantified variables, I don't know if it needs some extension enabled.. but since I'm having problems with the `Data.List.Extra` maybe it's my setup that's not right. Could you try for yourself and check if `coerce G.empty` works fine?

snowleopard replied:
Sorry, apparently things are a bit more complicated than I thought. See https://github.com/snowleopard/alga/pull/220#issuecomment-510891564.

bolt12 replied:
Thanks for your help! It's all good now!

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,907 @@
+-- 'size'        (vertex x) == 1
+-- @
+vertex :: a -> Graph a
+vertex = coerce . G.vertex
```
I think it should be `coerce G.vertex` -- does it work?

bolt12 replied:
Yes, it works fine, GHC actually gave me a type error with `coerce G.vertex`, that's why I changed it.

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,907 @@
+-- 'vertexCount' (edge 1 2) == 2
+-- @
+edge :: a -> a -> Graph a
+edge x y = connect (vertex x) (vertex y)
```
And this should be `edge = coerce G.edge`, etc.

snowleopard commented:
@bolt12 Thanks! I've left some more comments.

Not sure about the `Data.List.Extra` error.

By the way, could you please copy the text of errors instead of using screenshots? Screenshots are sometimes difficult to view, e.g. from a mobile.

bolt12 commented:
@snowleopard Sure!

```Haskell
[28 of 45] Compiling Algebra.Graph.Test ( test/Algebra/Graph/Test.hs, interpreted )

test/Algebra/Graph/Test.hs:24:1: error:
    Could not find module ‘Data.List.Extra’
    Use -v to see a list of the files searched for.
   |
24 | import Data.List.Extra (nubOrd)
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Failed, 27 modules loaded.
/bin/sh: cls: command not found
```

As I said, maybe my setup has some kind of misconfiguration... if I try to load a specific module it compiles just fine!

This is the `coerce G.empty` error:

```Haskell
src/Algebra/Graph/Undirected.hs:263:9: error:
    • Couldn't match representation of type ‘a0’ with that of ‘a’
        arising from a use of ‘coerce’
      ‘a’ is a rigid type variable bound by
        the type signature for:
          empty :: forall a. Graph a
        at src/Algebra/Graph/Undirected.hs:262:1-16
    • In the expression: coerce G.empty
      In an equation for ‘empty’: empty = coerce G.empty
    • Relevant bindings include
        empty :: Graph a (bound at src/Algebra/Graph/Undirected.hs:263:1)
    |
263 | empty = coerce G.empty
    |         ^^^^^^^^^^^^^^
```

snowleopard commented:
I'm not entirely sure what's going on with `coerce` in this case but, apparently, GHC just needs some help with type inference. For example, the following compiles:

```haskell
empty :: Graph a
empty = (coerce :: G.Graph a -> Graph a) G.empty

vertex :: a -> Graph a
vertex = (coerce :: (a -> G.Graph a) -> (a -> Graph a)) G.vertex
```

To make the code more readable, I suggest we add a couple of helper functions:

```haskell
coerce0 :: G.Graph a -> Graph a 
coerce0 = coerce

coerce1 :: (a -> G.Graph b) -> (a -> Graph b)
coerce1 = coerce

empty :: Graph a
empty = coerce0 G.empty

vertex :: a -> Graph a
vertex = coerce1 G.vertex
```

bolt12 commented:
@snowleopard thanks for your help on the `coerce`problem!

I've fixed some of the reviews you made. There's this function that I'm not sure if can be optimized somehow: `hasEdge`. Should I just use the the help from the `SymmetricRelation` implementation?

I still need to write the test suite and catch some possible bugs.

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,926 @@
+-----------------------------------------------------------------------------
+module Algebra.Graph.Undirected (
+    -- * Basic graph construction primitives
+    empty, vertex, edge, overlay, connect, vertices, edges, overlays, connects,
```
You need to export the data type `Graph`, as otherwise, it's impossible to use the library :) Just don't export the constructor `UG`.

bolt12 replied:
Ups, my bad!

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,926 @@
+-- 'edgeCount'   (overlay 1 2) == 0
+-- @
+overlay :: Graph a -> Graph a -> Graph a
+overlay (UG g1) (UG g2) = coerce (G.Overlay g1 g2)
```
Should be just `coerce2 G.overlay` I think.

snowleopard replied:
Same for some other functions below.

bolt12 replied:
I thought so too but had to define and auxiliar function to help GHC:

```Haskell
-- Help GHC with type inference (direct use of 'coerce' does not
-- compile).
coerce3 :: (Coercible a b, Coercible c d) => (a -> c -> G.Graph e) -> (b -> d -> Graph e)
coerce3 = coerce
```

This one can be used instead of `coerce2`!

snowleopard replied:
Aha, I see -- perhaps, you should just make `coerce2` more general using your implementation. By `coerceN` above I meant to name specialised `coerce` functions with N arguments.

bolt12 replied:
Right now I have 5 coerce type signatures, should I put them somewhere else or can I keep them in the file as of now?

snowleopard replied:
Keep them here for now.

bolt12 replied:
Okay!

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,926 @@
+
+-- Help GHC with type inference (direct use of 'coerce' does not
+-- compile).
+coerce2 :: (a -> a -> G.Graph b) -> (a -> a -> Graph b)
```
You can make this type signature a bit more general: `(a -> b -> G.Graph c) -> (a -> b -> Graph c)`.

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,926 @@
+-- edges [(x,y)]     == 'edge' x y
+-- edges [(x,y), (y,x)] == 'edge' x y
+-- @
+edges :: Eq a => [(a, a)] -> Graph a
```
I don't think you need to do anything special here, just use `coerce1 G.edges`.

snowleopard replied:
This whole module is mostly just one big `coerce` :)

You'll only need to be careful about deconstructing the graph -- you'll need to treat `Connect` as a commutative operation.

bolt12 replied:
I see that now! I'm going to change everywhere I can!

bolt12 replied:
I'm following the approach where `Connect a b` means that 'a' is connected to 'b' and 'b' is connected to 'a'. I could write `connect` like this:
```Haskell
connect :: Graph a -> Graph a -> Graph a
connect = coerce2 G.connect . coerce2 (flip G.connect)
```
But I don't think this would be a good idea! But with this approach I would need to filter the duplicate edges on the ' edges' function. What would you prefer?

the `edges`function would be like this:

```Haskell
edges :: Eq a => [(a, a)] -> Graph a
edges = coerce1 G.edges . nubBy (\x y -> x == swap y)
```

`nubBy` can be optimized!

snowleopard replied:
> But with this approach I would need to filter the duplicate edges on the `edges` function.

Why? I don't understand. You don't need to filter out anything.

bolt12 replied:
`edges` in the Graph module is defined like this `edges = overlays . map (uncurry edge)` where `edge` is just `connect (vertex x) (vertex y)` if, for the `Undirected` module I use the same implementation of `edges` then for `edges [(1,2),(2,1)] == Overlay (Connect (Vertex 1) (Vertex 2)) (Connect (Vertex 2) (Vertex 1))`, right?

If we don't want to represent the redundant edges in the Graph structure itself I think we need to filter the redundant edges (that's what I meant with duplicate)

snowleopard replied:
> If we don't want to represent the redundant edges in the Graph structure itself

This was never our goal. It would be very hard to guarantee that we never have redundant vertices or edges in an expression. If you look at `Algebra.Graph`, for example, users can also call `edges` with repeated edges, but we never filter them out -- it would complicate the implementation.

bolt12 replied:
You are right it would be very hard! I will change my mindset! I was doing this thinking that we would never have redundant edges and that's not the case! Thanks for snapping me out of it

bolt12 replied:
With this in mind I think it will be a little hard to keep the API exactly the same as the `SymmetricRelation` one.. for example: `edges [(x,y), (y,x)] == 'edge' x y` this property does not hold.

Should I keep the semantics of the undirected graph and relax some of the properties that hold with `SymmetricRelation`?

snowleopard replied:
> EDIT: Well it actually holds because the Show,Eq, etc.. instances convert it to a symmetric relation.

Exactly. We never expose the internal implementation, which allows us to have some redundant parts in expressions.

snowleopard replied:
>  Should this property hold: edgeCount (edges [(1,2),(2,1)]) == 1 ?

Yes, of course, there is only one undirected edge in this graph.

bolt12 replied:
Okay! Thanks for your help!

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,926 @@
+{-# NOINLINE [1] isSubgraphOf #-}
+{-# RULES "isSubgraphOf/Int" isSubgraphOf = isSubgraphOfIntR #-}
+
+-- Like 'isSubgraphOf' but specialised for graphs with vertices of type 'Int'.
```
This isn't currently specialised, so this comment is wrong. Please add a TODO to fix this.

bolt12 replied:
Yes, you're right! I forgot this ones too!

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,926 @@
+-- hasEdge x y                  == 'elem' (min x y, max x y) . 'edgeList'
+-- @
+hasEdge :: Eq a => a -> a -> Graph a -> Bool
+hasEdge s t (UG g) = hit g == Edge || hit' g == Edge
```
I suggest to simplify this to `G.hasEdge s t g || G.hasEdge t s g` for now, and add a TODO to optimise this further.

bolt12 replied:
Sounds good!

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,926 @@
+{-# INLINE [1] edgeSet #-}
+{-# RULES "edgeSet/Int" edgeSet = edgeIntSetR #-}
+
+-- Like 'edgeSet' but specialised for graphs with vertices of type 'Int'.
```
This isn't specialised.

bolt12 replied:
Fixed!

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,926 @@
+toSymmetricRelation :: Ord a => Graph a -> SR.Relation a
+toSymmetricRelation = foldg SR.empty SR.vertex SR.overlay SR.connect
+
+-- TODO: Change this to a specialised version when
```
Just delete this code for now.

bolt12 replied:
Alright!

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,926 @@
+-- path . 'reverse' == path 
+-- @
+path :: Eq a => [a] -> Graph a
+path xs = case xs of []     -> empty
```
Keep `coerce`ing!

snowleopard commented:
@bolt12 I left some more comments. Please don't stop with `coerce` -- almost all functions in this module are going to be just `coerce`! :)

bolt12 commented:
@snowleopard here (line 594 - 609)
```Haskell
-- | The sorted list of vertices of a given graph.
-- Complexity: /O(s * log(n))/ time and /O(n)/ memory.
--
-- @
-- vertexList 'empty'      == []
-- vertexList ('vertex' x) == [x]
-- vertexList . 'vertices' == 'Data.List.nub' . 'Data.List.sort'
-- @
vertexList :: Ord a => Graph a -> [a]
vertexList = Set.toAscList . vertexSet
{-# INLINE [1] vertexList #-}
{-# RULES "vertexList/Int" vertexList = vertexIntListR #-}

-- Like 'vertexList' but specialised for graphs with vertices of type 'Int'.
vertexIntListR :: Graph Int -> [Int]
vertexIntListR = IntSet.toList . vertexIntSetR
```

If I use `coerce G.vertexList` do I need to keep the specialised version?

snowleopard commented:
@bolt12 Yes, keep `Int`-specialised versions, especially if we can properly specialise them!

bolt12 commented:
@snowleopard Okay! I also realized that some of the specializations can benefit from the 'AdjacencyMap' module if we do not represent the redundant edges on the graph structure!

bolt12 commented:
@snowleopard Thank you very much for your quick reviews! I've pretty much coerced everything that made sense and fixed the `edges` function to respect the properties of an undirected graph! I've cleaned up the code too!

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,870 @@
+-- 'Algebra.Graph.Class.Undirected' type class, including the commutativity
+-- of 'connect'.
+-- For graphs that are known to be /non-empty/ at compile time, see
+-- "Algebra.Graph.NonEmpty". 'Graph' is an instance of type classes defined in
```
The comment about non-empty graphs is out-of-date unless you plan to implement non-empty undirected graphs too. I suggest to just delete it for now.

bolt12 replied:
Yes, my bad!

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,870 @@
+-- of 'connect'.
+-- For graphs that are known to be /non-empty/ at compile time, see
+-- "Algebra.Graph.NonEmpty". 'Graph' is an instance of type classes defined in
+-- modules "Algebra.Graph.Class" and "Algebra.Graph.HigherKinded.Class", which
```
These instances are currently missing.

bolt12 replied:
I plan on implementing them and the `Undirected` instance too!

bolt12 replied:
In what consists implementing the type classes defined in 'Algebra.Graph.HigherKinded.Class' ?

snowleopard replied:
No need to implement `Algebra.Graph.HigherKinded.Class`, so please remove it from the comment.

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,870 @@
+--
+-- This module defines and abstraction over the core data type 'Graph' and associated algorithms.
+-- Undirected graphs satisfy all laws of the
+-- 'Algebra.Graph.Class.Undirected' type class, including the commutativity
```
Why not add this instance?

bolt12 replied:
Yes, I am going to do that!

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,870 @@
+
+-- Help GHC with type inference (direct use of 'coerce' does not
+-- compile).
+coerce3 :: (b -> (a -> b) -> (b -> b -> b) -> (b -> b -> b) -> G.Graph a -> b) 
```
Let's call this `coerceFold`.

snowleopard replied:
Actually, since you use it only once, I think it shouldn't be a separate function. Just add a type signature locally.

bolt12 replied:
Okay!

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,870 @@
+-- connects [x]       == x
+-- connects [x,y]     == 'connect' x y
+-- connects           == 'foldr' 'connect' 'empty'
+-- 'isEmpty' . connects == 'all' 'isEmpty'
```
Please migrate all extra examples you have in `Symmetric.Relation` to this module, together with the associated tests. For example, in this case we have this extra example: `connects == connects . reverse`.

bolt12 replied:
It must have failed me to migrate some examples. I'll pay attention to these examples.

bolt12 commented:
@snowleopard I've added some functions that were missing from the Symmetric Relation API and implemented the missing type classes with the exception of the HigherKinded because I didn't understand well it's objective. Could you please explain it to me!

Once again, thanks for your reviews!

bolt12 commented:
@snowleopard I started to add the test-suite for this new module. I needed to add the `toGraph` projection for the 'UndirectedGraph' type and export it since we do not export the constructor ( this was needed to get the test-suite working) as well as the `toSymmetricRelation` which I think it's no problem.

There are a couple of problems with the test-suite mainly because there's no UndirectedAdjacencyMap implementation and I need to work around it!

Please tell me what you think :)

snowleopard commented:
> with the exception of the HigherKinded because I didn't understand well it's objective. Could you please explain it to me!

@bolt12 No need to add an instance for higher-kinded graphs for now. I'm keen to remove this type class at some point anyway.

Could you clarify what exactly is the problem with the testsuite? Please give an example that doesn't work.

bolt12 commented:
@snowleopard Sure!

Here is the error when I run the test suite:

```Haskell
AdjacencyMap

Test failure:
    toAdjacencyMap             == foldg empty vertex overlay connect

*** Failed! Falsified (after 3 tests):
edge 1 2

*** Exception: ExitFailure 1
```

As you can see from the `Undirected` `ToGraph` type class instance: `toAdjacencyMap = toAdjacencyMap . UG.toSymmetricRelation`, this way, although it is slow, we do not need an implementation of `UndirectedAdjacencyMap` (yet) because we convert the `Undirected` Graph into a symmetric relation! The problem with the test suite is that it's testing `toAdjacencyMap == foldg AM.empty AM.vertex AM.overlay AM.connect ` which is not true!

I'm reusing the same testsuite used for `SymmetricRelation` since they should pass all tests but these little detail will make me create a different testsuite just for these case where I'll check that `toAdjacencyMap == foldg AM.empty AM.vertex AM.overlay AM.connect . toSymmetricRelation` (and in the future `toAdjacencyMap == foldg UAM.empty UAM.vertex UAM.overlay UAM.connect`)


Does creating a different test function `testUndirectedToGrap` makes sense? should I work around it differently?

snowleopard commented:
@bolt12 Thanks, I see. The problem is that `foldg` expects the arguments to satisfy the axioms of undirected graphs, but `AM.connect` doesn't satisfy it -- it's not commutative.

Perhaps, we shouldn't implement `ToGraph` at all. Can you drop this instance?

bolt12 commented:
@snowleopard Sure! Should I leave a TODO note to add the instance once a possible `UndirectedAdjacencyMap` is implemented?

snowleopard commented:
> Should I leave a TODO note to add the instance once a possible `UndirectedAdjacencyMap` is implemented?

No, no need for it. I think we'll actually need a new `ToGraph` type class for undirected graphs.

Please keep only the functions exported from the new module in the API.

bolt12 commented:
@snowleopard Okay! I think everything is ready to change this PR to a normal one and see what are the problems with the CI checks! :)

snowleopard commented on test/Main.hs:
```diff
@@ -27,17 +28,18 @@ main :: IO ()
 main = do
     selected <- getArgs
     let go current = when (null selected || current `elem` selected)
-    go "Acyclic.AdjacencyMap"   testAcyclicAdjacencyMap
```
Please bring back the dots.

bolt12 replied:
Ups, I don't know how those disappeared! Everything is fixed now!

snowleopard commented:
> Edit: It was just a merge conflict, I hope everything is fine now!

Not quite, I think you resolved the conflict incorrectly.

bolt12 commented:
@snowleopard Yes, I'm sorry, my bad! Hopefully there's nothing broken anymore.

snowleopard commented on test/Algebra/Graph/Test/Undirected.hs:
```diff
@@ -0,0 +1,44 @@
+-- Stability  : experimental
+--
+-- Testsuite for "Algebra.Graph.Undirected" and polymorphic functions defined in
+-- "Algebra.Graph.HigherKinded.Class".
```
I'm not sure the "and polymorphic functions..." part should be here.

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,913 @@
+-----------------------------------------------------------------------------
+module Algebra.Graph.Undirected (
+    -- * Algebraic data type for graphs
+    Graph, toGraph,
```
I think we should rename `toGraph` to `fromUndirected` as in `Algebra.Graph.Relation.Symmetric`.

bolt12 replied:
Makes sense! I will change that :)

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,913 @@
+import qualified Algebra.Graph                 as G
+import Algebra.Graph.Internal 
+
+import qualified Algebra.Graph.Relation.Symmetric    as SR
```
Alignment.

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,913 @@
+    > 1 + 2       == Overlay (Vertex 1) (Vertex 2)
+    > 1 * 2       == Connect (Vertex 1) (Vertex 2)
+    > 1 + 2 * 3   == Overlay (Vertex 1) (Connect (Vertex 2) (Vertex 3))
+    > 1 * (2 + 3) == Connect (Vertex 1) (Overlay (Vertex 2) (Vertex 3))
```
The above list is no longer correct since constructors are not exported -- please use the expressions that are printed by the `Show` instance.

bolt12 replied:
Yes! I forgot to change this particular text when I changed the show intance! Thanks for pointing it out!

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,913 @@
+instance (Show a, Ord a) => Show (Graph a) where
+  show = show . toSymmetricRelation
+
+{- See Note [Functions for rewrite rules] in 'Algebra.Graph' -}
```
@nobrakal Do you think we need to use any rewrite rules in this module? We simply wrap `Graph` into a `newtype` and most of the functions below are defined simply via `coerce`. I have a feeling that the rewrite rules from your module may be sufficient, in which case this module could be simplified by removing all rewrite rules. What are your thoughts?

nobrakal replied:
Indeed, you do not need to use rules. Express `bindR` using `G.Graph`'s `bind`.
Also inline everything (or at least, remove (now) useless phase control for `bindR`, `foldg` `overlays`...), since we need to show to the compiler that fusion is possible !

snowleopard replied:
@nobrakal Thanks! 

@bolt12 Could you follow Alexandre's suggestions?

bolt12 replied:
@snowleopard Sure! :-) Just to make sure: I need to remove the rules, rewrite bind, and inline everything, right? I didn't understand the remove (now) part very well @nobrakal

nobrakal replied:
Sorry if my comment was not clear. We need to:

* Remove all rules from `Algebra.Graph.Undirected` concerning functions that are just synonym of functions from `Algebra.Graph` (since their rules are already living in `Algebra.Graph`).
* Remove phase control (the [X] part in `INLINE` pragma) for functions that are just synonym of functions from `Algebra.Graph`. They are not needed anymore.
* The goal is to use rules from `Algebra.Graph`, thus we need to express all possible functions using functions from `Algebra.Graph`. This apply particularly to `bindR`.

This should (I hope) make things work :)

bolt12 replied:
Yes, thank you very much!

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,913 @@
+-- removeEdge 1 2 (1 * 1 * 2 * 2)  == 1 * 1 + 2 * 2
+-- @
+removeEdge :: Eq a => a -> a -> Graph a -> Graph a
+removeEdge s t = coerce (G.removeEdge s t . G.removeEdge t s) 
```
What does `coerce` do in this case? I think it should be removed.

bolt12 replied:
`coerce` makes the `G.removeEdge s t . G.removeEdge t s` which is of type `G.Graph a -> G.Graph a`, to be of type `Graph a -> Graph a`, right?

I can change the function and not use `coerce` with explicit pattern matching!

bolt12 replied:
@snowleopard forgot to mention you in the comment above :-)

snowleopard replied:
Thanks, I think I got it, let's keep it as is.

P.S.: No need to mention me, I get notifications either way :) It's just a very busy week.

bolt12 replied:
Thanks, and sorry I forgot it was a review! I didn't mean to be impatient! Take your time!

snowleopard commented:
@bolt12 I made another pass and left some minor comments. I think we're getting ready to merge, thank you!

bolt12 commented:
@snowleopard, @nobrakal I think I did what I need to! Since I don't have experience with ghc inlining pragmas I dont know if I inlined something I shouldn't or if I removed the correct phase control... I could use some help to see if everything is alright!

Thanks for your help!

nobrakal commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,919 @@
+{- See Note [Functions for rewrite rules] in 'Algebra.Graph' -}
+
+instance Functor Graph where
+    fmap = fmapR
```
```suggestion
    fmap f = UG . fmap f . fromUndirected
    {-# INLINE fmap #-}
```

nobrakal commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,919 @@
+
+instance Applicative Graph where
+    pure  = vertex
+    (<*>) = apR
```
```suggestion
    (<*>) = coerce3 (<*>)
    {-# INLINE (<*>) #-}
```

nobrakal commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,919 @@
+
+instance Monad Graph where
+    return = pure
+    (>>=)  = bindR
```
```suggestion
    (>>=) g = UG . (>>=) (fromUndirected g) . coerce
    {-# INLINE (>>=) #-}
```

nobrakal commented:
Hi @bolt12, you get the idea :) I have added a couple of suggestions to use even more code from `Algebra.Graph`.

This should do the trick, some tests will need to be made after #228 to ensure that all is well!

bolt12 commented:
@nobrakal, Thanks for your suggestions! I've applied them and all the tests pass as of now, hope everything is alright when your PR is merged :-D

snowleopard commented on test/Algebra/Graph/Test/API.hs:
```diff
@@ -292,6 +293,51 @@ graphAPI = API
     , compose                    = G.compose
     , box                        = G.box }
 
+-- | The API of 'UG.Graph.
```
Please check all generated docs: the closing quote is missing here.

bolt12 replied:
Thanks! I've checked all the generated docs and fixed some small issues like this, I hope it's all good now!

snowleopard commented on test/Main.hs:
```diff
@@ -33,6 +34,7 @@ main = do
     go "Bipartite.AdjacencyMap" testBipartiteAdjacencyMap
     go "Export"                 testExport
     go "Graph"                  testGraph
+    go "Undirected"             testUndirected
```
Please keep this list ordered alphabetically.

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,911 @@
+
+        >   x * empty == x
+        >   empty * x == x
+        >   x * y     == y * x
```
```suggestion
        >       x * y == y * x
```

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,911 @@
+{-# RULES "eqR/Int" eqR = eqIntR #-}
+
+-- Like 'eqR' but specialised for graphs with vertices of type 'Int'.
+-- NOTE: This is not specialised to vertices of type 'Int'. But it's still
```
```suggestion
-- TODO: This is not specialised to vertices of type 'Int'. But it's still
```

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,911 @@
+{-# RULES "ordR/Int" ordR = ordIntR #-}
+
+-- Like 'ordR' but specialised for graphs with vertices of type 'Int'.
+-- NOTE: This is not specialised to vertices of type 'Int'. But it's still
```
```suggestion
-- TODO: This is currently not specialised to vertices of type 'Int'. But it's still
```

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,911 @@
+{-# RULES "isSubgraphOf/Int" isSubgraphOf = isSubgraphOfIntR #-}
+
+-- Like 'isSubgraphOf' but specialised for graphs with vertices of type 'Int'.
+-- NOTE: This is not specialised to vertices of type 'Int'. But it's still
```
```suggestion
-- TODO: This is currently not specialised to vertices of type 'Int'. But it's still
```

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,911 @@
+{-# RULES "vertexCount/Int" vertexCount = vertexIntCountR #-}
+
+-- Like 'vertexCount' but specialised for graphs with vertices of type 'Int'.
+-- NOTE: This is not specialised to vertices of type 'Int'. But it's still
```
```suggestion
-- TODO: This is currently not specialised to vertices of type 'Int'. But it's still
```

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,911 @@
+{-# RULES "edgeCount/Int" edgeCount = edgeCountIntR #-}
+
+-- Like 'edgeCount' but specialised for graphs with vertices of type 'Int'.
+-- NOTE: This is not specialised to vertices of type 'Int'. But it's still
```
```suggestion
-- TODO: This is currently not specialised to vertices of type 'Int'. But it's still
```

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,911 @@
+{-# RULES "edgeList/Int" edgeList = edgeIntListR #-}
+
+-- Like 'edgeList' but specialised for graphs with vertices of type 'Int'.
+-- NOTE: This is not specialised to vertices of type 'Int'. But it's still
```
```suggestion
-- TODO: This is currently not specialised to vertices of type 'Int'. But it's still
```

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -0,0 +1,911 @@
+{-# RULES "edgeSet/Int" edgeSet = edgeIntSetR #-}
+
+-- Like 'edgeSet' but specialised for graphs with vertices of type 'Int'.
+-- NOTE: This is not specialised to vertices of type 'Int'. But it's still
```
```suggestion
-- TODO: This is currently not specialised to vertices of type 'Int'. But it's still
```

snowleopard commented:
@bolt12 Apologies for taking so long to review and merge. I've added some more suggestions.

bolt12 commented:
@snowleopard no problem at all, this month is a little busy for me too!

I resolved your reviews and gave a look at the generated docs. I think it's all good! Let me know if there's still anything I should correct!

Thanks for your reviews!

snowleopard commented on test/Main.hs:
```diff
@@ -1,20 +1,21 @@
+import           Algebra.Graph.Test.Relation
+import           Algebra.Graph.Test.Relation.SymmetricRelation
+import           Algebra.Graph.Test.Undirected
+import           Data.Graph.Test.Typed
```
These changes seem accidental, please undo them. Same for the two imports below.

bolt12 replied:
It is! I'm sorry about this, my Haskell setup was doing this automatically. I already disabled it!

snowleopard commented on test/Algebra/Graph/Test/Undirected.hs:
```diff
@@ -0,0 +1,43 @@
+testUndirected = do
+    putStrLn "\n============ Graph.Undirected ============"
+    test "Axioms of undirected graphs" $
+        size10 (undirectedAxioms :: GraphTestsuite G)
```
Compared to `SymmetricRelation`, we miss some tests:

https://github.com/snowleopard/alga/blob/master/test/Algebra/Graph/Test/Relation/SymmetricRelation.hs

For example, `fromUndirected` is not tested. I guess we need to test it. 

Do we need to add `toUndirected` too, for consistency of the API?

bolt12 replied:
On it! I think it makes sense to provide `toUndirected` it wont hurt since it will be just a coercion, right?

bolt12 replied:
Actually, I don't think that `toUndirected` and `fromUndirected` should work as `toSymmetric` and  `fromSymmetric` since we decided to convert it to a normal `Graph` without redundant edges. Some of the properties will be different, is that okay?

snowleopard commented:
@bolt12 Thanks, I've just left a couple of more comments :-)

snowleopard commented:
@bolt12 I think `fromUndirected` is currently wrong: it allows you to observe the difference of `edge 1 2` and `edge 2 1`. I believe we should make the symmetric closure before allowing to introspect the structure of an `Undirected` graph.

snowleopard commented on src/Algebra/Graph/Undirected.hs:
```diff
@@ -179,7 +180,7 @@ instance (Show a, Ord a) => Show (Graph a) where
 
 instance Functor Graph where
-    fmap f = UG . fmap f . fromUndirected
+    fmap f = UG . fmap f . _fromUndirected
```
I think we probably don't need `_fromUndirected` at all, and should use `coerce` here and in `>>=`.

bolt12 replied:
You're right!

snowleopard commented:
@bolt12 Many thanks for the PR and all the fixes. Merged now! 

It's very nice that we now have undirected algebraic graphs! I'll make a pass with minor revisions at some point but your implementation and tests look great already.

bolt12 commented:
@snowleopard Once again, it was my pleasure! Thank you for your time and reviews, I hope everything is right on point!
I want to thank @nobrakal too, for the help!
